Master adf4371 support

stefpopa opened the pull request:
The ADF4371 is a frequency synthesizer with an integrated voltage
controlled oscillator (VCO) for phase-locked loops (PLLs). The ADF4371
has an integrated VCO with a fundamental output frequency ranging from
4000 MHz to 8000 MHz. In addition, the VCO frequency is connected to
divide by 1, 2, 4, 8, 16, 32, or 64 circuits that allows the user to
generate radio frequency (RF) output frequencies as low as 62.5 MHz at
RF8x. A frequency multiplier at RF16x generates from 8 GHz to 16 GHz. A
frequency quadrupler generates frequencies from 16 GHz to 32 GHz at RF32x.
RFAUX8x duplicates the frequency range of RF8x or permits direct access to
the VCO output.

The driver takes the reference input frequency from the device tree and
uses it to calculate and maximize the PFD frequency (frequency of the phase
frequency detector). The PFD frequency is further used to calculate the
timeouts: synthesizer lock, VCO band selection, automatic level
calibration (ALC) and PLL settling time.

This initial driver exposes the attributes for setting the frequency and
enabling/disabling the different adf4371 channels.

Datasheet:
Link: https://www.analog.com/media/en/technical-documentation/data-sheets/adf4371.pdf

nunojsa commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+	unsigned int readval, reg, bit;
+	int ret;
+
+	mutex_lock(&indio_dev->mlock);
```
From a previous review, i got the feeling that is not really advised to use `mlock`. Shouldn't we initialize our own mutex here?

stefpopa replied:
Yes, you are right! Good catch!

nunojsa commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+		return ret;
+
+	st->clkin_freq = clk_get_rate(st->clkin);
+
```
Dumb question... Is it actually a real scenario that the `clkin` frequency changes after we set  `st->clkin_freq`?

stefpopa replied:
`clkin `does not change, but we are also using it outside probe in `adf4371_clk_disable()` when we have to call `clk_disable_unprepare(st->clkin);`

nunojsa replied:
Yes, i got that. I was just wondering if we could get a stall value in `st->clkin`...

stefpopa commented:
@mhennerich: this driver is as basic as it gets. Other features will follow.

stefpopa commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+	do_div(val, ADF4371_MODULUS1 * (1 << ref_div_sel));
+
+	if (channel == ADF4371_CH_RF16)
+		val <<= 1;
```
I wouldn't create a macro for this. It is a plain multiplication with 2 for the frequency doubler channel and a multiplication with 4 for the frequency quadrupler channel. A macro would be overkill.

commodo replied:
i agree with Stefan here; macro is a bit too much for this; 
this is fine as-is

stefpopa commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+	}
+	mutex_unlock(&indio_dev->mlock);
+
+	return ret < 0 ? ret : sprintf(buf, "%llu\n", val);
```
You are right that "upstream people" are not huge fans of the ternary operator. However, there is no rule that doesn't allow it, as long at is obvious what it does. 
You will often find it used in situations just like this one.

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+#define ADF4371_REG71	0x71
+#define ADF4371_REG72	0x72
+#define ADF4371_REG73	0x73
+#define ADF4371_REG7C	0x7C
```
I'm not sure defining these here provide much value;
maybe if they would have reg-names

stefpopa replied:
I don't feel strongly about this either. Are you suggesting we should get rid of them altogether and do stuff like `regmap_write(st->regmap, 0x00, ADF4371_RESET_CMD) `instead of `regmap_write(st->regmap, ADF4371_REG0, ADF4371_RESET_CMD)`  ?

commodo replied:
maybe just add the registers (with reg-names) that are used with their names, and don't define the others
see https://github.com/analogdevicesinc/linux/pull/375/commits/fb66d1b100dfc16b2e612bfafb1698b52a45c74c#r283664468

commodo replied:
oh; 
since these are the official names in the datasheet, let's add a comment about it here and do a parametrized macro:
`#define ADF4371_REG(x)  (0x # x)`  or something like that

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+#define ADF4371_REG7C	0x7C
+
+/* ADF4371_REG0 */
+#define ADF4371_ADDR_ASC_MSK		BIT(2)
```
I guess for ADF4371_REG0; it makes a bit more sense to do 
`#define ADF4371_ADDR  0x00`

stefpopa commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+		if (ret < 0)
+			break;
+
+		readval &= ~BIT(bit);
```
It would be great, but I couldn't figure out how to use it in a clean manner in this specific case. The main issue is that you cannot pass complex macros that take a parameter to structs.

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+	*mod2 /= gcd_div;
+	*fract2 /= gcd_div;
+
+	return 0;
```
if not using the return value, maybe make this function void

stefpopa replied:
makes sense

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+	switch (channel) {
+	case ADF4371_CH_RF8:
+	case ADF4371_CH_RFAUX8:
+		if ((freq > ADF4371_MAX_OUT_FREQ) ||
```
just for fun, these look like they could be made into a fun macro like;
```
#define ADF4371_CHECK_RANGE(freq, range) \ 
    if ((freq > ADF4371_MAX_ #range) ||
         (freq < ADF4371_MIN_ #range))
          return -EINVAL;
```

and call with   `ADF4371_CHECK_RANGE(freq, OUT_FREQ)`

commodo replied:
you could also leave it like this;

stefpopa replied:
:smile: it looks fun indeed! However, I am not sure if this kind of macro wouldn't fall under the category of "macros that affect control flow": https://www.kernel.org/doc/html/v4.10/process/coding-style.html#macros-enums-and-rtl

commodo replied:
ack; you can also just use the macro to do range-check evaluation;
so
```
#define ADF4371_CHECK_RANGE(freq, range) \ 
    ((freq > ADF4371_MAX_ #range) ||
         (freq < ADF4371_MIN_ #range))
```

stefpopa replied:
:+1:

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+		break;
+	default:
+		ret = -EINVAL;
+		val = 0;
```
maybe add a `break` here  to avoid GCC from complaining

stefpopa replied:
ack

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+			break;
+
+		readval &= ~BIT(bit);
+		readval |= (!!!readin << bit);
```
why 3 `!!!` ? 
isn't 1 sufficient ?

stefpopa replied:
You are probably right. The compiler will optimize it as single `!` anyway

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+		ret = regmap_write(st->regmap, reg, readval);
+		break;
+	default:
+		ret = -EINVAL;
```
`break` for GCC

stefpopa replied:
ack

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,633 @@
+			timeout = 2;
+			synth_timeout++;
+		}
+	} while (synth_timeout * 1024 + timeout <= 20 * tmp);
```
maybe move these while loops in some inline helpers;

stefpopa replied:
From my understanding, the use of inline functions in the Linux kernel is not encouraged. See [the inline disease](https://www.kernel.org/doc/html/v4.10/process/coding-style.html#the-inline-disease).

commodo replied:
The arguments in that document don't apply much here.
In this case inline is more to get a neater code, while abstracting some parts in some functions.
There's a sentence there `While the use of inlines can be appropriate (for example as a means of replacing macros, see Chapter 12), it very often is not. ` ; I'm guessing that would refer (mostly) to inline functions that are added in header files, and have more than 3 lines. 

In this case, inline doesn't do much to change the generated code; mostly it does with tidying it up a bit.
We can also leave it like this here.
Not feeling strongly about moving these to functions or not.
Let's leave it as-is

stefpopa commented:
Changes in v2:
* defined a `ADF4371_REG(x) (0x ## x) `macro instead of a separate macro for each register.
* defined a `ADF4371_CHECK_RANGE(freq, range)`
* added a `mutex `struct as part of `adf4371_state`
* changed the type of `adf4371_pll_fract_n_compute()` to `void`
* added `break `statements where needed

commodo commented on drivers/iio/frequency/adf4371.c:
```diff
@@ -0,0 +1,573 @@
+	struct spi_device *spi;
+	struct regmap *regmap;
+	struct clk *clkin;
+	struct mutex mlock;
```
nitpick: renamed  `mlock` -> `lock`
to avoid any confusions with indio_dev's mlock

stefpopa commented:
changes in v3:
* renamed `mlock `-> `lock`

commodo commented:
Let's leave this PR open till monday.
Maybe we'll get some answer from upstream.

DT has been acked; let's see about the driver.

stefpopa commented:
changes in v4 are following the remarks received during upstream review:
* Added a new sysfs-bus-iio-frequency-adf4371 file which documents the ABI changes.
* Modified the ADF4371_REG() macro to take the reg values in hex as params
* ADF4371_MAX_MODULUS2 macro is now defined as BIT(14)
* regmap_bulk_write() can't do DMA directly, so the buffer was forced into it's own cacheline.
* Fixed the multi line comment style.

stefpopa commented:
changes in v5 following the remarks received during second round of upstream review:
 - out_altvoltageY_frequency and out_altvoltageY_powerdown attributes are treated as normal indexed channels
- out_altvoltageY_name attribute was added, from which the datasheet names of the channels can be read.
- Added more information in the documentation.
- Documented the use of mutex lock.
- As part of adf4371_write(), used a bool variable for power down and a 64 bit variable for the frequency.
